docs/projects/2024/architecture-transition/rpc-client.md

```diff
@@ -0,0 +1,153 @@
+# JSON-RPC client plans
+
+## Examples of how the API would be called
+
+### Non-batched, non-cancellable
+
+```ts
+/** @type {ApiRequest<Spacecraft>>} */
+const request = api.spacecrafts.get({ id: 42 });
+
+/** @type {Promise<Spacecraft>} */
+const promise = request.run();
+
+/** @type {Spacecraft} */
+const spacecraft = await promise; // Might throw!
+```
+
+### Non-batched, cancellable
+
+```ts
+/** @type {ApiRequest<Spacecraft>>} */
+const request = api.spacecrafts.get({ id: 42 });
+
+/** @type {CancellablePromise<Spacecraft | undefined>} */
+const promise = request.cancellableRun();
+
+/**
+ * Resolves to `undefined` if canceled.
```

Comment:
Cancelled requests should not resolve or reject at higher layers. They definitely should not resolve to `undefined`.

The idea behind a cancelled request is that any layer of top of it should not have to worry about it anymore. Resolving to `undefined` makes it a problem of the higher layers instead of dealing with it within CancellablePromise.